webapp/views/App/views/Data/Charts/Charts.js

```diff
@@ -1,150 +1,124 @@
-  // const [showStackedLegent, setShowStackedLegend] = useState(false)
+const useChartSpec = () => {
+  const [spec, setSpec] = useState(BASE_SPEC)
+
+  const updateSpecRaw = useCallback((newSpec) => {
+    setSpec(newSpec)
+  }, [])
+
+  const updateSpec = useCallback((newSpec) => {
+    updateSpecRaw(A.parse(newSpec))
+  }, [])
 
-  const survey = useSurvey()
+  return { spec, updateSpec }
+}
+
+const useChartRender = (spec, query) => {
   const surveyId = useSurveyId()
   const cycle = useSurveyCycleKey()
 
-  useEffect(() => {
-    const getChart = async () => {
-      try {
-        if (dimensionDefUuids.length <= 0) return
-        const { data } = await axios.post(getUrl({ surveyId, query: Query.create({ entityDefUuid }) }), {
-          cycle,
-          query: A.stringify(Query.create({ entityDefUuid })),
-          chart: A.stringify(buildSpec({ dimensions: dimensionDefUuids, survey })),
-        })
-        setChartImage(data.svg)
-      } catch (err) {
-        console.log(err)
-      }
+  const [chartImage, setChartImage] = useState(null)
+
+  const renderChart = useCallback(async () => {
+    try {
+      if (!query) return
+      const { data } = await axios.post(getUrl({ surveyId, query }), {
+        cycle,
+        query: A.stringify(query),
+        chart: A.stringify(spec),
+      })
+
+      setChartImage(data.svg)
+    } catch (err) {
+      console.log(err)
     }
-    getChart()
-    return () => {}
-  }, [surveyId, cycle, entityDefUuid, dimensionDefUuids])
+  }, [spec, surveyId, cycle])
+
+  return {
+    chartImage,
+    renderChart,
+  }
+}
+
+const useChart = (query) => {
+  const [draft, setDraft] = useState(null)
+
+  const { spec, updateSpec } = useChartSpec()
+  const { chartImage, renderChart } = useChartRender(spec, query)
+
+  useEffect(() => {
+    setDraft(false)
+  }, [chartImage])
+
+  useEffect(() => {
+    setDraft(true)
+  }, [spec])
+
+  return { spec, updateSpec, draft, chartImage, renderChart }
+}
+const dimensions = [
+  {
+    name: 'city_label',
+    value: 'city_label',
+    label: 'City',
+    type: 'nominal',
+  },
+  {
+    name: 'inhabitants',
+    value: 'inhabitants',
+    label: 'Inhabitants',
+    type: 'quantitative',
+  },
+]
+
+const Charts = () => {
+  const [entityDefUuid, setEntityDefUuid] = useState(null)
+
+  const { spec, updateSpec, chartImage } = useChart(entityDefUuid ? Query.create({ entityDefUuid }) : null)
 
   return (
     <div className="charts">
-      <Chart src={chartImage ? `data:image/svg+xml;base64,${btoa(unescape(encodeURIComponent(chartImage)))}` : false} />
+      <DataSelector setEntityDefUuid={setEntityDefUuid} entityDefUuid={entityDefUuid} />
+
       <Panel
         setEntityDefUuid={setEntityDefUuid}
         entityDefUuid={entityDefUuid}
-        setDimensionDefUuids={setDimensionDefUuids}
-        dimensionDefUuids={dimensionDefUuids}
+        spec={spec}
+        onUpdateSpec={updateSpec}
+        dimensions={dimensions}
```

Comment:
the Panel component is empty now, why do you need to pass these props to it?

Reply:
Yep, you are right. We are going to need them but not at the moment.